build_tools/circle/build_test_arm.sh

```diff
@@ -4,57 +4,82 @@ set -e
     for name in gcc g++ cc c++ x86_64-linux-gnu-gcc x86_64-linux-gnu-c++; do
       ln -s $(which ccache) "/tmp/ccache/${name}"
+      which ${name}
     done
-    export PATH="/tmp/ccache/:${PATH}"
-    ccache -M 256M
+    # Unset ccache limits
+    ccache -F 0
+    ccache -M 0
 }
 
 # imports get_dep
 source build_tools/shared.sh
 
 sudo add-apt-repository --remove ppa:ubuntu-toolchain-r/test
 sudo apt-get update
-sudo apt-get install python3-virtualenv ccache
-python3 -m virtualenv --system-site-packages --python=python3 testenv
-source testenv/bin/activate
-pip install --upgrade pip
+
+# Setup conda environment
+MINICONDA_URL="https://github.com/conda-forge/miniforge/releases/latest/download/Mambaforge-Linux-aarch64.sh"
+
+# Install Mambaforge
+wget $MINICONDA_URL -O mambaforge.sh
+MINICONDA_PATH=$HOME/miniconda
+chmod +x mambaforge.sh && ./mambaforge.sh -b -p $MINICONDA_PATH
+export PATH=$MINICONDA_PATH/bin:$PATH
+mamba init --all --verbose
+mamba update --yes conda
+
+# Create environment and install dependencies
+mamba create -n testenv --yes $(get_dep python $PYTHON_VERSION)
+source activate testenv
+
+# Use the latest by default
+mamba install --verbose -y  ccache \
+                            pip \
+                            $(get_dep numpy $NUMPY_VERSION) \
+                            $(get_dep scipy $SCIPY_VERSION) \
+                            $(get_dep cython $CYTHON_VERSION) \
+                            $(get_dep joblib $JOBLIB_VERSION) \
+                            $(get_dep threadpoolctl $THREADPOOLCTL_VERSION) \
+                            $(get_dep pytest $PYTEST_VERSION) \
+                            $(get_dep pytest-xdist $PYTEST_XDIST_VERSION)
 setup_ccache
-python -m pip install $(get_dep cython $CYTHON_VERSION) \
-                      $(get_dep joblib $JOBLIB_VERSION)
-python -m pip install $(get_dep threadpoolctl $THREADPOOLCTL_VERSION) \
-                      $(get_dep pytest $PYTEST_VERSION) \
-                      $(get_dep pytest-xdist $PYTEST_XDIST_VERSION)
 
 if [[ "$COVERAGE" == "true" ]]; then
-    python -m pip install codecov pytest-cov
-fi
-
-if [[ "$PYTEST_XDIST_VERSION" != "none" ]]; then
-    python -m pip install pytest-xdist
+    mamba install --verbose -y codecov pytest-cov
 fi
 
 if [[ "$TEST_DOCSTRINGS" == "true" ]]; then
     # numpydoc requires sphinx
-    python -m pip install sphinx
-    python -m pip install numpydoc
+    mamba install --verbose -y sphinx
+    mamba install --verbose -y numpydoc
 fi
 
 python --version
+mamba list
 
-# Set parallelism to 3 to overlap IO bound tasks with CPU bound tasks on CI
-# workers with 2 cores when building the compiled extensions of scikit-learn.
-export SKLEARN_BUILD_PARALLEL=3
+# Set parallelism to $N_CORES + 1 to overlap IO bound tasks with CPU bound tasks on CI
+# workers with $N_CORES cores when building the compiled extensions of scikit-learn.
+export SKLEARN_BUILD_PARALLEL=$(($N_CORES + 1))
+pip install --verbose .
```

Comment:
The default build isolation used by `pip` will numpy again in a virtualenv with a temporary directory that changes between runs. This means that ccache will get cache misses when the compiler includes files from `numpy`. I think this may work:

```
pip install --verbose --no-build-isolation .
```

Or if that does not work, I expect this would:

```
python setup.py install
```

Reply:
Also [local project installs](https://pip.pypa.io/en/stable/cli/pip_install/?highlight=%20Local%20project%20installs#local-project-installs) pip will copy the entire project directory to a temporary one. I think we can add `--use-feature=in-tree-build` to the `pip` command to keep the cache warm for source files in sklearn:

```python
pip install --verbose --no-build-isolation --use-feature=in-tree-build .
```